src/core/algorithms/od/fastod/complex_stripped_partition.h

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "attribute_set.h"
+#include "range_based_stripped_partition.h"
+
+namespace algos::fastod {
+
+class ComplexStrippedPartition {
+private:
+    std::shared_ptr<std::vector<size_t>> sp_indexes_;
+    std::shared_ptr<std::vector<size_t>> sp_begins_;
+    std::shared_ptr<std::vector<DataFrame::range_t>> rb_indexes_;
+    std::shared_ptr<std::vector<size_t>> rb_begins_;
+    std::shared_ptr<DataFrame> data_;
+    bool is_stripped_partition_;
+    bool should_be_converted_to_sp_;
+
+    static constexpr inline double SMALL_RANGES_RATIO_TO_CONVERT = 0.5;
+    static constexpr inline size_t MIN_MEANINGFUL_RANGE_SIZE = static_cast<size_t>(40);
+
+    std::string sp_ToString() const;
+    void sp_Product(short attribute);
+    bool sp_Split(short right) const;
+
+    std::string rb_ToString() const;
+    void rb_Product(short attribute);
+    bool rb_Split(short right) const;
+
+    std::vector<DataFrame::value_indexes_t> IntersectWithAttribute(
+            algos::fastod::AttributeSet::size_type attribute, size_t group_start, size_t group_end);
+
+    ComplexStrippedPartition(std::shared_ptr<DataFrame> data,
+                             std::shared_ptr<std::vector<size_t>> indexes,
+                             std::shared_ptr<std::vector<size_t>> begins);
+
+    ComplexStrippedPartition(std::shared_ptr<DataFrame> data,
+                             std::shared_ptr<std::vector<DataFrame::range_t>> indexes,
+                             std::shared_ptr<std::vector<size_t>> begins);
+
+public:
+    ComplexStrippedPartition();
+    ComplexStrippedPartition(ComplexStrippedPartition const& origin) = default;
+
+    ComplexStrippedPartition& operator=(ComplexStrippedPartition const& other);
+
+    std::string ToString() const;
+    void Product(short attribute);
+    bool Split(short right) const;
+
+    bool ShouldBeConvertedToStrippedPartition() const;
+    void ToStrippedPartition();
+
+private:
+    template <bool ascending>
+    bool sp_Swap(short left, short right) const {
```

Comment:
Seems that `sp_Swap` and `rb_Swap` differs only in `values` filling, maybe we could merge them into one function with additional bool argument like `is_stripped_partition_`?

Reply:
Did it.